traversal/planner/GraphPlanner.java

```diff
@@ -231,37 +233,43 @@ Optimiser optimiser() {
     }
 
     void mayOptimise(GraphManager graphMgr, boolean singleUse) {
-        if (procedure == null) {
-            try {
-                firstOptimisingLock.writeLock().lock();
-                if (procedure == null) optimise(graphMgr, singleUse);
-                assert procedure != null;
-            } finally {
-                firstOptimisingLock.writeLock().unlock();
-            }
-        } else if (isOptimising.compareAndSet(false, true)) {
-            try {
-                optimise(graphMgr, singleUse);
-            } finally {
-                isOptimising.set(false);
-            }
+        long timeLimitMillis = singleUse ? HIGHER_TIME_LIMIT_MILLIS : DEFAULT_TIME_LIMIT_MILLIS;
+        if (backgroundOptimisation == null) beginFirstOptimise(graphMgr, timeLimitMillis);
+        else if (isOptimising.compareAndSet(false, true)) beginReOptimise(graphMgr, timeLimitMillis);
+
+        try {
+            backgroundOptimisation.get(timeLimitMillis, MILLISECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException ignored) {
+            LOG.trace("Query plan did not finish updating in " + timeLimitMillis + " ms.");
+        }
+    }
+
+    private synchronized void beginFirstOptimise(GraphManager graphMgr, long timeLimitMillis) {
+        if (backgroundOptimisation == null) {
+            isOptimising.set(true);
+            updateTraversalCosts(graphMgr);
+            updateOptimiser();
+            createProcedure();
+            backgroundOptimisation = CompletableFuture.runAsync(() -> optimise(timeLimitMillis), async2());
         }
     }
 
-    private void optimise(GraphManager graphMgr, boolean singleUse) {
+    private void beginReOptimise(GraphManager graphMgr, long timeLimitMillis) {
         updateTraversalCosts(graphMgr);
         if (isUpToDate() && isOptimal()) {
             if (LOG.isDebugEnabled()) LOG.debug("GraphPlanner still optimal and up-to-date");
+            isOptimising.set(false);
             return;
         }
         if (!isUpToDate()) updateOptimiser();
 
-        // TODO: we should have a more clever logic to allocate extra time
-        long allocatedDuration = singleUse ? HIGHER_TIME_LIMIT_MILLIS : DEFAULT_TIME_LIMIT_MILLIS;
-        Instant start, endSolver, end;
+        backgroundOptimisation = backgroundOptimisation.thenRunAsync(() -> optimise(timeLimitMillis), async2());
+    }
 
+    private void optimise(long timeLimitMillis) {
+        Instant start, endSolver, end;
         start = Instant.now();
-        optimiser.optimise(allocatedDuration);
+        optimiser.optimise(timeLimitMillis);
```

Comment:
the computation of the costs and `isUpToDate`-ness is checked before calling `optimise()` now. This method is purely responsible for running the optimiser, setting the state to `isUpToDate = true`, creating the new procedure, and finishing with `isOptimising = false`.

We could argue that we should set the `isOptimising = false` outside of this method, but inside the future's runnable.